### PR TITLE
[SERVICE-581] Toast manager hardening

### PR DESCRIPTION
**Pull request by pjbroadbent**
Updates to ToastManager/Layouter/etc to make toasts and toast animations more robust. Can handle multiple notifications being created simultaneously, and the re-creation of a toast due to id re-use. Includes additional safety around toast creation in particular, including destroying a toast before the toast has been fully created and initialised.

There are also a number of changes to the demo app windows, to facilitate in the manual testing of these edge cases. The default size of these windows has been updated due to the additional buttons, the windows may look weird if using saved window state.

Changelist:
- Added LayoutStack class for storing Toast objects and enforcing Toast constraints.
- Modified Toast initialisation.
- Removed some methods from ToastManager that were very similar to each other.
- Added clarifying comments around code that handles edge cases and race conditions.
- Fin window wrapper ensures no calls are made to the underlying window objects once the window has closed.
- Added buttons to demo app to spawn/clear multiple notifications simultaneously.
- When raising the same notification ID from a demo app, titles will now be different.
- Added recently added provider components to the provider window object.
- Launcher app will exit if all other demo app windows are closed.

### CODE REVIEW

File: src/demo/launcher.ts [hunk 1]
```diff
@@ -5,22 +5,29 @@ const baseUrl: string = window.location.href.split('/').slice(0, -1).join('/');
...
-        }, console.error);
+        const app = await fin.Application.start(conf);
+
+        promises.push(new Promise(resolve => {
+            app.once('closed', resolve);
+        }));
     }
+
+    // If all of the apps launched above close, then close this app too
+    Promise.all(promises).then(() => fin.Application.getCurrentSync().close());
```
[the remark this answers is not in the document]
Comment (pjbroadbent) [reply src/demo/launcher.ts]: I tried the latter approach, but seems there's an RVM race condition that causes the service to exit with the launcher app.

Didn't spend too much time looking into it, just went with this approach instead. Good catch with the restart case, will update to an `isRunning`-style check.

[the remark this answers is not in the document]
Comment (tomrobertsOF) [reply src/demo/launcher.ts]: Fair enough on the first point. As I said, not really a big deal either way.

File: src/provider/controller/Layouter.ts [hunk 2]
```diff
@@ -112,27 +108,36 @@ export class Layouter extends AsyncInit {
     }
 
     /**
-     * Layout a given stack of Layoutable items
+     * Layout a given stack of Layoutable items. Toast positions are calculated and assigned synchronously.
```
[the remark this answers is not in the document]
Comment (pjbroadbent) [reply src/provider/controller/Layouter.ts]: I was kinda looking at this the other way round - as in, does it have a need for the abstraction? I definitely left this as a weird mix of `Toast` and `LayoutItem`, but I'm leaning more towards removing the interface completely than to re-introduce it into `LayoutStack`.

File: src/provider/controller/Layouter.ts [hunk 3]
```diff
@@ -112,27 +108,36 @@ export class Layouter extends AsyncInit {
...
-            const newPosition: PointTopLeft = {
-                top: (this.direction <= 0) ? top : prev,
-                left: left
-            };
-            item.position = newPosition;
-            this.moveItem(item, item.position);
-            // update stack height.
-            stack.layoutHeight += height + Layouter.INTERNAL_CONFIG.spacing;
+            const dimensions = item.dimensions;
+
```
[the remark this answers is not in the document]
Comment (pjbroadbent) [reply src/provider/controller/Layouter.ts]: I don't see any overlap when running the demo app, but this is in need of some clean-up. It's a little inconsistent around where various offsets are applied.

**Pull request merged**